checks/tasks/rpki.py

```diff
@@ -0,0 +1,435 @@
+    base=SetupUnboundContext)
+def mail_rpki(self, mx_ips_pairs, url, *args, **kwargs):
+    """Celery task to perform rpki test on mailservers for a domain."""
+    return do_mail_rpki(mx_ips_pairs, url, self, *args, **kwargs)
+
+
+@mail_registered
+@shared_task(
+    bind=True,
+    soft_time_limit=settings.SHARED_TASK_SOFT_TIME_LIMIT_HIGH,
+    time_limit=settings.SHARED_TASK_TIME_LIMIT_HIGH,
+    base=SetupUnboundContext)
+def mail_mx_ns_rpki(self, mx_ips_pairs, url, *args, **kwargs):
+    """Celery task to perform rpki test on nameservers for the mx records of a domain."""
+    return do_mx_ns_rpki(mx_ips_pairs, url, self, *args, **kwargs)
+
+
+@batch_mail_registered
+@shared_task(
+    bind=True,
+    soft_time_limit=settings.BATCH_SHARED_TASK_SOFT_TIME_LIMIT_HIGH,
+    time_limit=settings.BATCH_SHARED_TASK_TIME_LIMIT_HIGH,
+    base=SetupUnboundContext)
+def batch_mail_rpki(self, mx_ips_pairs, url, *args, **kwargs):
+    """Celery task to perform rpki test on mailservers for a domain."""
+    return do_mail_rpki(mx_ips_pairs, url, self, *args, **kwargs)
+
+
+@batch_mail_registered
+@shared_task(
+    bind=True,
+    soft_time_limit=settings.BATCH_SHARED_TASK_SOFT_TIME_LIMIT_HIGH,
+    time_limit=settings.BATCH_SHARED_TASK_TIME_LIMIT_HIGH,
+    base=SetupUnboundContext)
+def batch_mail_ns_rpki(self, mx_ips_pairs, url, *args, **kwargs):
+    """Celery task to perform rpki test on nameservers for a domain.
+
+    `@mail_registered` passes in mx_ips_pairs, which are not needed for this test.
+    """
+    return do_ns_rpki(url, self, *args, **kwargs)
+
+
+@batch_mail_registered
+@shared_task(
+    bind=True,
+    soft_time_limit=settings.BATCH_SHARED_TASK_SOFT_TIME_LIMIT_HIGH,
+    time_limit=settings.BATCH_SHARED_TASK_TIME_LIMIT_HIGH,
+    base=SetupUnboundContext)
+def batch_mail_mx_ns_rpki(self, mx_ips_pairs, url, *args, **kwargs):
+    """Celery task to perform rpki test on nameservers for the mx records of a domain."""
+    return do_mx_ns_rpki(mx_ips_pairs, url, self, *args, **kwargs)
+
+
+def report_exists(subtestname, category, domainset) -> None:
+    """Generate a test report for the existence of ROAs."""
+    def gen_tech_data(domain, ip, validity) -> List[List[str]]:
+        # Provide tech_data to generate a table of the following form
+        #
+        # Server     | IP address  | ROA exists
+        # -------------------------------------
+        # example.nl | 192.168.0.1 | yes
+        # example.nl | 192.168.0.2 | no
+
+        if any(route['vrps'] for route in validity.values()):
+            row = [domain, ip, "detail tech data yes"]
+        else:
+            row = [domain, ip, "detail tech data no"]
+
+        return row
+
+    failure_count = 0
+    tech_data = []
+
+    prev_domain = None
+    for domain in domainset:
+        for ip in domain.routing:
+            # failure to validate, team cymru or routinator was unavailable
+            if not ip['routes'] or not all(ip['validity'].values()):
```

Comment:
This needs more thought. Currently #655 and addresses that Team Cymru can't find routes for are treated as a local error. That's not right.